### ISSUE TITLE
How do I handle Vercel preview and Supabase auth?

### ISSUE THREAD

**Issue by activenode** [supabase/supabase#4808]
Hey there. I think this one is massively important and I am unsure why it wasn't an issue before.

Nowadays most Continouous Deployments consist of automated deploys. Let it be on render.com on vercel.com or what not. Many of those create a preview URL such as `some-hash.provider.ext` .

Now supabase expects a "generic" base URL - which I would like to understand why because from an architectural perspective I think there should be only a list of "allowed URLs" and not necessarily a primary URL as the "primary" term is kinda not helpful (at least I am not seeing the advantage from an outside perspective).

Now since the Auth configuration in the UI is very static (input(Main URL), input(additional URLs)) there is no way of having a proper automated workflow - at all. There will always be manual work involved of adding more and more URLs.

IMO this can be even considered as "architectural flaw/bug".

I'd be happy to help here if you point into the directions that need to be adapted. Other than that I'd be grateful if someone could pick that topic up, it is massively time-consuming and I am having a hard time understanding that this wasn't a problem for other people so far.

![image](https://user-images.githubusercontent.com/9058762/148199368-a1c9be4a-cbd0-445a-a1d9-d698ad97b0df.png)

### Discussed in https://github.com/supabase/supabase/discussions/2760

<div type='discussions-op-text'>

<sup>Originally posted by **pedrobslisboa** August  7, 2021</sup>
Hey guys, I've been looking for the answer to this question without success.

### How do I handle Vercel preview and Supabase auth?
E.g.: How can I have the redirect URL for auth with google provider and sync to Vercel?

<img width="1452" alt="Captura de Tela 2021-08-07 às 09 09 29" src="https://user-images.githubusercontent.com/35539594/128599666-849c104c-09ee-40bd-9a89-0d8d4ea0e454.png">
</div>

**Comment by activenode** [comment]
I might be looking into this myself soon cause its so crucial but I cannot promise to get it fixed. Will check ASAP

**Comment by MildTomato** [comment]
Tagged @awalias and @kangmingtay as there's some discussion here about using wildcards as the base URL as a possible solution to supporting multiple preview URLs from vercel/netlify.

**Comment by activenode** [comment]
Okay so one thing I found out so far is that _theoretically_ the supabase repo doesn't need any adaption. It would be nice from a UI perspective but technically supabase so far seems to only save the strings.

Now the actual check is performed within `gotrue` . See here  https://github.com/activenode/gotrue/blob/8ba49df7c7bbcc87c9d5b211a32cb5aa17db231d/api/helpers.go#L127

**Comment by activenode** [comment]
```
for _, uri := range config.URIAllowList {
		if redirectURL == uri {
			return true
		}
	}
```

Do you see any problem with adapting this to match Wildcards? I mean it would be a major change. We could instead introduce flags but those flags would need to be mapped from supabase forward to gotrue (e.g. `switch-on wildcards`) which I find kinda unnecessary to maintain.

I'd either go for 

1. Breaking Change: The For Loop checks for wildcard patterns
2. Adding yet another field: `config.URIWildcards` but again: I find it kinda cumbersome

Besides that, as @MildTomato mentioned I am even thinking with shrinking the overall code to just "ONE" url -> `URIAllowList`. Now if for whatever reason you would need a fallback "root" url you could say that the first one in the allow list will always be the fallback/root one.

Thoughts? @awalias @kangmingtay

**Comment by milanvanschaik** [comment]
@activenode I've made a PR to get it started: https://github.com/supabase/gotrue/pull/334

**Comment by activenode** [comment]
I'd personally opt for the change done by @milanvanschaik  (as stated in my previous comments) but we would definitely need supabase to clarify on the potential concerns.

**Comment by kangmingtay** [comment]
Hi @milanvanschaik and @activenode, apologies for the late reply, the main reason why we don't accept wildcards or arbitrary paths in the redirect url list is due to [this reason](https://github.com/supabase/gotrue/issues/123#issuecomment-871214361). 

As for the workaround, let me discuss with the team further on how we can decouple the mobile deeplinks from the standard redirect urls and get back to you.

**Comment by milanvanschaik** [comment]
@kangmingtay Maybe I'm overlooking something, but wouldn't it be sufficient to either:
1) Advice against configure redirect URIs with wildcard for mobile redirect paths, or:
2) Validate on this - should start with http(s) protocol for wildcard usage?

**Comment by multiplehats** [comment]
We're really bumping into this issue with testing RLS and other auth features on Vercel preview URLs. There's really no workaround :/

**Comment by activenode** [comment]
@kangmingtay . From an architectural point of view, even though you added the related discussion/reason I am still having troubles understanding this. Problem-bound.

Like what will happen if we simply change it? In the related comment it says that it will be hard to match mobile links but then again how so? Like what will happen and what use cases do we have? Explicit ones.

Like e.g. 

I think stating those here would help progressing this topic. Because the stated comment 
> may not conform to the standard scheme://sub.domain.tld pattern hence why we can not match the host securely with wildcard paths etc.

including _may_ implies that there are no actual known problems yet. Why not work from "Help WebDevs first and then help solving problems on mobile when they appear" ?

**Comment by awalias** [comment]
related: https://github.com/supabase/gotrue/issues/99

**Comment by awalias** [comment]
also an alternative approach: https://github.com/supabase/gotrue/issues/296

**Comment by qruz-hq** [comment]
Couldn't Supabase "just" implement different CLIENT API keys and tie them to different domains ? 

The only real solution for now is running parallel databases but we run too much data to duplicate it anyways.

**Comment by obetomuniz** [comment]
Ralated https://github.com/supabase/supabase/issues/5658

**Comment by benallfree** [comment]
Jumping in here to mention the problem is the same on Cloudflare Pages.

**Comment by simonh1000** [comment]
Agreed, and here is an example of a netlify preview url https://6224790c45550f0007015694--my-site.netlify.app/
I want to be able to redirect to urls like this during development

**Comment by mrpbennett** [comment]
Is there no away around this? I am using Vercel for a company prototype with a supabase backend, using Google auth which is needed due to the nature of the prototype.

**Comment by kaspnilsson** [comment]
@mrpbennett what's worked for me as a workaround:

- Create a few Git branches that are meant specifically for pending pull requests (`dev`, `dev2` are mine)
- Use a `NEXT_PUBLIC_SUPABASE_REDIRECT_URL` env variable to set your redirect URL in app
- Set up your environment variables in Vercel targeting these branches specifically with corresponding redirect urls
- When working on a given feature, do development largely locally
- When hosting PRs for review, only use `dev` and `dev2` => `main`

**Comment by mrpbennett** [comment]
Thanks for the tip @kaspnilsson  would this work if I am using Vite and not NextJS

**Comment by kaspnilsson** [comment]
I've never used Vite, but if you still can use environment variables and Vercel then I'd imagine the same strategy could work!

**Comment by henrycunh** [comment]
If we could add redirect URLs through a REST API this would be quite simple to automate, too...

**Comment by henrycunh** [comment]
> @henrycunh so you're saying we can use a vercel hosted domain with google auth redirect now? 

for sure :]
just add a valid glob pattern. it wasn't mentioned ANYWHERE but in a tweet from someone that doesn't work at supabase, but i'm glad it works!

**Comment by henrycunh** [comment]
https://twitter.com/kevcodez/status/1523649173790535680

**Comment by activenode** [comment]
Can this be officially confirmed @awalias ? Then we could close this issue here <3

**Comment by kangmingtay** [comment]
Hey guys, apologies we missed this in the previous monthly update but you can now handle vercel previews with a wildcard in your redirect url. We follow the glob patterns listed [here](https://pkg.go.dev/github.com/gobwas/glob#Compile). We also take into account "." and "/" as separators so `*.example.com` would match on `api.example.com` but not `foo.api.example.com`. Using the "**" pattern ignores the separators but I would recommend trying to keep the glob pattern as strict as possible. 

We're also in the midst of updating the auth settings page on the dashboard to reflect this change so stay tuned!

**Comment by aantn** [comment]
Hi, thank you for adding this. It doesn't fully solve the problem for Netlify, as the DNS for Netlify previews is `some_id--my_org.netlify.app`. To support this, we would need a pattern like `*--my_org.netlify.app`

**Comment by awalias** [comment]
@aantn could you try `**--my_org.netlify.app` according to this is should work

![image](https://user-images.githubusercontent.com/458736/172354887-ddffc2cb-42d6-4ec9-84bd-86884c46b447.png)

**Comment by kiwicopple** [comment]
Hey everyone, closing this now that we have the wildcard matching. Thanks for the patience!

**Comment by travispamaral** [comment]
I know this is closed now but Im hoping someone (@kiwicopple or @activenode) can chime in. I have followed IDK how many issues and PRs and random tweets but for the life of me I CAN NOT get the redirect url to work between localhost and vercel preview deployments. Based on a few examples I tried a `getURL()` function that checks NODE_ENV and I can log out the response with I try to auth with google and I see the right url (vercel) but then on redirect it just goes localhost.

- Should my **SITE_URL** be the prod and not localhost?
- Which fields are the proper fields to put my `localhost:3000`, `app-*-123xyz.vercel.app` urls in the supabase settings?

**Comment by kiwicopple** [comment]
sorry to hear you're having difficulty!

- `SITE_URL` = make this your official site URL
- Additional site URLs = should be something like this: `http://localhost:3000/*/*,https://*vercel.app/*/*`

Then when you do a signIn you can do:

```js
const { user, session, error } = await supabase.auth.signIn({
  provider: 'github'
}, {
  redirectTo: process.env.NEXT_PUBLIC_SITE_URL // set this to your site URL as an env var
})
```

**Comment by thorwebdev** [comment]
I've added a section in the docs which will hopefully help others: https://supabase.com/docs/guides/auth#redirect-urls-and-wildcards

**Comment by dalmo3** [comment]
I struggled to get localhost working after following the documentation. 

I set `emailRedirectTo: window.location.href`, which resolves to `http://localhost:5137/`.

I first tried adding `http://localhost:5137/` as redirect URL, then tried `http://localhost:5137/**` - just like the example from the docs, but neither of those worked.

What worked for me was `http://localhost:**` - for some reason the pattern matcher doesn't like the port? In any case, just leaving a comment as it might help someone else.
